### ISSUE TITLE
“Serverless Error: Rest API id could not be resolved.” but Serverless service does not use Rest API resources

### ISSUE THREAD

**Issue by carchi1a** [serverless/serverless#9286]
Experiencing an intermittent issue when running serverless deploy. Configuration does not use API Gateway or any API Gateway-associated resources; however, I get the following error message at the end of my deployment: 

> Rest API id could not be resolved.
> 
> This might be caused by a custom API Gateway configuration.
> 
> In given setup stage specific options such as `tracing`, `logs` and `tags` are not supported.
> 
> Please update your configuration (or open up an issue if you feel that there's a way to support your setup).

Executing the command a (variable) number of times eventually results in a successful deployment.

<details>
<summary><code>serverless.yml</code></summary>

```yaml
service: utils

plugins:
  - serverless-plugin-ifelse
  - serverless-webpack
  - serverless-vpc-discovery
  - serverless-deployment-bucket
  - serverless-pseudo-parameters
  - serverless-prune-plugin
  - serverless-stack-termination-protection
  - serverless-plugin-lambda-insights

custom:
  prefix: aicpa-mxp-lambdas
  
  stage: dummy

  versions_enabled:
    local: false
    sit: true
    uat: true
    staging: true
    pt: true
    preview: true
    prod: true

  serverless_deployment_s3_bucket_name:
    example: ${self:service.prefix}-example

  vpc:
    vpcName: aicpa-member-experience-${self:provider.stage}
    subnetNames:
      - Private_1a-1_${self:provider.stage}
      - Private_1b-1_${self:provider.stage}
    securityGroupNames:
      - serverless-lambda

  currentStage: ${self:provider.stage}
  
  serverlessIfElse:
    - If: '"${self:custom.currentStage}" == "local"'
      Exclude:
        - functions.example1
    - If: '"${opt:custom_stage, self:custom.stage}" == "example"'
      Set:
        - custom.lambdaInsights.defaultLambdaInsights: false

  prune:
    automatic: true
    includeLayers: true
    number: 5

  serverlessTerminationProtection:
    stages:
      - prod

  lambdaInsights:
    defaultLambdaInsights: true
    attachPolicy: false

provider:
  name: aws
  stackName: ${self:custom.prefix}-${opt:stage}-${self:service.name}
  runtime: nodejs10.x
  stage: ${opt:custom_stage, opt:stage}
  tracing:
    lambda: PassThrough
  timeout: 30
  logRetentionInDays: 180
  deploymentBucket:
    name: ${self:custom.serverless_deployment_s3_bucket_name.${opt:stage, 'sit'}, self:custom.serverless_deployment_s3_bucket_name.sit}
    serverSideEncryption: AES256 # Server-side encryption method
    blockPublicAccess: true # Prevents public access via ACLs or bucket policies
  versionFunctions: true
  environment:
    STAGE: ${opt:stage, 'sit'}
    NODE_ENV: production
    ENVIRONMENT: ${self:provider.stage}
    SUPPRESS_NO_CONFIG_WARNING: true
    ALLOW_CONFIG_MUTATIONS: true
    
functions:
  example1:
    role: arn:aws:iam::#{AWS::AccountId}:role/${self:custom.prefix}-${self:provider.stage}-example1-role
    name: ${self:custom.prefix}-${opt:stage}-v3-example1
    handler: handler-utils.example1
    memorySize: ${ssm:/mxp-plus/${self:provider.stage}/example1_MEMORY~true}
   
  example2:
    role: arn:aws:iam::#{AWS::AccountId}:role/${self:custom.prefix}-${self:provider.stage}-example2-role
    name: ${self:custom.prefix}-${opt:stage}-v3-example2
    handler: handler-utils.example2
    
```

</details>

<details>
<summary><b><code>node --max-old-space-size=4096 node_modules/serverless/bin/serverless deploy --custom_stage example --base_path rest-1126 --api_gateway_id_ssm_name random --api_gateway_root_resource_id_ssm_name random -s rest-1126 -c serverless_utils.yml</code> output</b></summary>

```
(node:26578) ExperimentalWarning: The fs.promises API is experimental

(node:26578) ExperimentalWarning: The dns.promises API is experimental

Serverless: Deprecation warning: "provider.stage" is not accessible (configured behind variables which cannot be resolved at this stage).

            Starting with next major release, this will be communicated with a thrown error.

            Set "variablesResolutionMode: 20210219" in your service config, to adapt to this behavior now

            More Info: https://www.serverless.com/framework/docs/deprecations/#NEW_VARIABLES_RESOLVER

Serverless: To ensure safe major version upgrades ensure "frameworkVersion" setting in service configuration (recommended setup: "frameworkVersion: ^2.31.0")



Serverless: Load command interactiveCli

Serverless: Load command config

Serverless: Load command config:credentials

Serverless: Load command config:tabcompletion

Serverless: Load command config:tabcompletion:install

Serverless: Load command config:tabcompletion:uninstall

Serverless: Load command create

Serverless: Load command install

Serverless: Load command package

Serverless: Load command deploy

Serverless: Load command deploy:function

Serverless: Load command deploy:list

Serverless: Load command deploy:list:functions

Serverless: Load command invoke

Serverless: Load command invoke:local

Serverless: Load command info

Serverless: Load command logs

Serverless: Load command metrics

Serverless: Load command print

Serverless: Load command remove

Serverless: Load command rollback

Serverless: Load command rollback:function

Serverless: Load command slstats

Serverless: Load command plugin

Serverless: Load command plugin

Serverless: Load command plugin:install

Serverless: Load command plugin

Serverless: Load command plugin:uninstall

Serverless: Load command plugin

Serverless: Load command plugin:list

Serverless: Load command plugin

Serverless: Load command plugin:search

Serverless: Load command config

Serverless: Load command config:credentials

Serverless: Load command upgrade

Serverless: Load command uninstall

Serverless: Load command webpack

Serverless: Load command prune

Serverless: Load command login

Serverless: Load command logout

Serverless: Load command generate-event

Serverless: Load command test

Serverless: Load command dashboard

Serverless: Load command output

Serverless: Load command output:get

Serverless: Load command output:list

Serverless: Load command param

Serverless: Load command param:get

Serverless: Load command param:list

Serverless: Load command studio

Serverless: Deprecation warning: Resolution of lambda version hashes was improved with better algorithm, which will be used in next major release.

            Switch to it now by setting "provider.lambdaHashingVersion" to "20201221"

            More Info: https://www.serverless.com/framework/docs/deprecations/#LAMBDA_HASHING_VERSION_V2

Serverless: Invoke deploy

Serverless: Invoke package

Serverless: Invoke aws:common:validate

Serverless: Using deployment bucket 'lambdas-example'

Serverless: Invoke aws:common:cleanupTempDir

Serverless: Updating VPC config...

Serverless: Invoke webpack:validate

Serverless: Invoke webpack:compile

Serverless: Bundling with Webpack...

Hash: bbbcf0cdeb7177ebdeca

Version: webpack 4.46.0

Time: 76869ms

Built at: 04/13/2021 2:08:38 PM

                                   Asset      Size  Chunks                   Chunk Names

config/custom-environment-variables.json  8.43 KiB          [emitted]        

                     config/default.json  7.07 KiB          [emitted]        

            config/development-mock.json  5.48 KiB          [emitted]        

                 config/development.json  7.77 KiB          [emitted]        

                        handler-utils.js  21.1 MiB       0  [emitted]        handler-utils

                    handler-utils.js.map  12.6 MiB       0  [emitted] [dev]  handler-utils

Entrypoint handler-utils = handler-utils.js handler-utils.js.map

   [0] ./node_modules/aws-sdk/lib/core.js 2.43 KiB {0} [built]

   [1] ./node_modules/aws-sdk/lib/node_loader.js 3.88 KiB {0} [built]

   [4] ./node_modules/tslib/tslib.es6.js 10 KiB {0} [built]

   [7] ./node_modules/mxp-service-common/dist/index.js 1.37 KiB {0} [built]

  [10] external "fs" 42 bytes {0} [built]

  [11] ./node_modules/mxp-schemas/dist/index.js 2.95 KiB {0} [built]

  [14] ./node_modules/mxp-service-common/dist/lib/logger.js 5.7 KiB {0} [built]

  [18] external "path" 42 bytes {0} [built]

  [31] ./node_modules/mxp-utils/dist/index.js 3.07 KiB {0} [built]

 [115] ./node_modules/@sentry/node/esm/index.js + 46 modules 175 KiB {0} [built]

       | ./node_modules/@sentry/node/esm/index.js 1.44 KiB [built]

       | ./node_modules/@sentry/node/esm/version.js 118 bytes [built]

       | ./node_modules/@sentry/node/esm/transports/index.js 155 bytes [built]

       | ./node_modules/@sentry/node/esm/backend.js 4.64 KiB [built]

       | ./node_modules/@sentry/node/esm/client.js 1.38 KiB [built]

       | ./node_modules/@sentry/core/esm/index.js 698 bytes [built]

       | ./node_modules/@sentry/node/esm/handlers.js 9.88 KiB [built]

       | ./node_modules/@sentry/node/esm/integrations/index.js 309 bytes [built]

       | ./node_modules/@sentry/node/esm/sdk.js 4.63 KiB [built]

       | ./node_modules/@sentry/types/esm/severity.js 1.4 KiB [built]

       | ./node_modules/@sentry/types/esm/status.js 1.44 KiB [built]

       | ./node_modules/@sentry/hub/esm/scope.js 14.5 KiB [built]

       | ./node_modules/@sentry/types/esm/session.js 384 bytes [built]

       | ./node_modules/@sentry/hub/esm/session.js 3.14 KiB [built]

       | ./node_modules/@sentry/hub/esm/hub.js 15.5 KiB [built]

       |     + 32 hidden modules

 [150] ./node_modules/aws-sdk/lib/aws.js 159 bytes {0} [built]

 [212] ./node_modules/csvtojson/v2/index.js 281 bytes {0} [built]

 [342] ./node_modules/source-map-support/register.js 25 bytes {0} [built]

 [836] ./node_modules/source-map-support/source-map-support.js 19.3 KiB {0} [built]

[3574] ./handler-utils.ts + 13 modules 97.5 KiB {0} [built]

       | ./handler-utils.ts 1.12 KiB [built]

       | ./services/init.ts 3.78 KiB [built]

       | ./handlers/utils/functions/example1.ts 4.32 KiB [built]

       | ./handlers/utils/functions/example2.ts 571 bytes [built]

       | ./handlers/helpers/dynamo-helper.ts 558 bytes [built]

       | ./handlers/helpers/commercetools.shims.ts 3.1 KiB [built]

       | ./handlers/helpers/topics.ts 7.14 KiB [built]

       | ./services/ElasticSearch.ts 20.8 KiB [built]

       | ./handlers/helpers/commerceTools.helpers.ts 25.3 KiB [built]

    + 3594 hidden modules

Serverless: Invoke webpack:package

Serverless: Packaging service...

Serverless: Invoke aws:package:finalize

AWS Pseudo Parameters

Skipping automatic replacement of regions with account region!

AWS Pseudo Parameter: Resources::example1::Properties::Role Replaced AWS::AccountId with ${AWS::AccountId}

AWS Pseudo Parameter: Resources::example2::Properties::Role Replaced AWS::AccountId with ${AWS::AccountId}

Serverless: Invoke aws:common:moveArtifactsToPackage

Serverless: Invoke aws:common:validate

Serverless: Using deployment bucket 'lambdas-example'

Serverless: Invoke aws:deploy:deploy

Serverless: Uploading CloudFormation file to S3...

Serverless: Uploading artifacts...

Serverless: Uploading service utils.zip file to S3 (7.15 MB)...

Serverless: Validating template...

Serverless: Creating Stack...

Serverless: Checking Stack create progress...

.................................................................

Serverless: Stack create finished...

Serverless: Invoke aws:info

Service Information

service: utils

stage: rest-1126

region: us-east-1

stack: lambdas-rest-1126-utils

resources: 21

api keys:

  None

endpoints:

  None

functions:

  example1: lambdas-example1

  example2: lambdas-example2

layers:

  None

 

 Serverless Error ----------------------------------------

 

  ServerlessError: Rest API id could not be resolved.

  This might be caused by a custom API Gateway configuration.

  

  In given setup stage specific options such as `tracing`, `logs` and `tags` are not supported.

  

  Please update your configuration (or open up an issue if you feel that there's a way to support your setup).

      at resolveAccountInfo.call.then.then.then (/home/jenkins/workspace/rest-1126/packages/lambdas/node_modules/serverless/lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.js:84:19)

      at process._tickCallback (internal/process/next_tick.js:68:7)

```

</details>

<b>Installed version</b>

```
Operating System:          linux

     Node Version:              10.16.3

     Framework Version:         2.31.0 (local)

     Plugin Version:            4.5.2

     SDK Version:               4.2.2

     Components Version:        3.7.7
```

**Comment by pgrzesik** [comment]
Hello @carchi1a, thanks for reporting. After looking at your config, it seems that the `provider.tracing` property is not configured properly - it's even highlighted with schema validation warning in the output. Could you please correct it and check if the issue still exists?

**Comment by carchi1a** [comment]
Hi @pgrzesik, thank you for looking into this so promptly. 

I had noticed the validation warning, but I had ignored since it was the required configuration for the `serverless-plugin-tracing`. I gather that this plugin is now obsolete as this functionality is built in, so following the Serverless documentation I updated my config to define tracing as:

```
tracing:
    lambda: PassThrough
```

I can see that the warning is no longer there; however, I still experience this issue intermittently. I've edited the original comment to reflect the new .yml and the output from running the same command. Any further help would be much appreciated. Thanks!

**Comment by pgrzesik** [comment]
Hello @carchi1a - thanks for followup. Thanks quite interesting, because in order to hit the error you're experiencing you would have to enable either tracing or logs for API Gateway. Are you able to test it out with reduced number of plugins? Maybe one of them is introducing that setting behind the scenes? Also, in your command, I've noticed `--api_gateway_id_ssm_name random --api_gateway_root_resource_id_ssm_name` flags, what are they used for?

**Comment by carchi1a** [comment]
Hi @pgrzesik, I've been running some tests and I have found a plausible explanation for the issue. The `--api_gateway_id_ssm_name random --api_gateway_root_resource_id_ssm_name` flags you asked about we use in our deployment script because we deploy 4 stacks in parallel (via Jenkins parallel). These flags are not applicable for this stack, but we use them in a different stack to force Serverless to create an API Gateway (in some cases), and we do enable logs for this stack. 

I believe this attempt to parallelise the deployment is what is causing an issue, and would explain why we experience the issue sometimes but not all the time - depending on where the process jumps from one thread to another and what the state of each stack's deployment is at. I've serialised the deployment of the 4 stacks and haven't noticed the issue since. I'm closing this issue, thanks again for looking into this.
